src/freetype_manager.cc

```diff
@@ -0,0 +1,596 @@
+{
+    if (!gFtFontsInitialized) {
+        return;
+    }
+
+    font -= 100;
+
+    if (gFtFontDescriptors[font].filebuffer != NULL) {
+        gCurrentFtFont = font;
+        current = &(gFtFontDescriptors[font]);
+    }
+}
+
+// 0x442168
+static int FtFontGetLineHeightImpl()
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->lineSpacing + current->maxHeight + current->heightOffset;
+}
+
+// 0x442188
+static int FtFontGetStringWidthImpl(const char* string)
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    if (strlen(string) == 1 && string[0] < 0) {
+        return current->wordSpacing;
+    } else {
+        int count;
+        int width = 0;
+
+        count = LtoU((char*)string, strlen(string));
+
+        for (int i = 0; i < count; i++) {
+            uint32_t ch = output[i];
+
+            if (ch == L'\n' || ch == L'\r') {
+                continue;
+            }
+
+            if (ch == '\x95') {
+                FtFontGlyph g = GetFtFontGlyph(ch);
+                width += g.width + current->letterSpacing + 2;
+            } else if (ch == L' ' || (ch < 256 && ch > 128)) {
+                width += current->wordSpacing + current->letterSpacing;
+            } else {
+                FtFontGlyph g = GetFtFontGlyph(ch);
+                width += g.width + current->letterSpacing;
+            }
+        }
+
+        return width;
+    }
+}
+
+// 0x4421DC
+static int FtFontGetCharacterWidthImpl(int ch)
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->wordSpacing;
+}
+
+// 0x442210
+static int FtFontGetMonospacedStringWidthImpl(const char* str)
+{
+    return FtFontGetStringWidthImpl(str);
+}
+
+// 0x442240
+static int FtFontGetLetterSpacingImpl()
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->letterSpacing;
+}
+
+// 0x442258
+static int FtFontGetBufferSizeImpl(const char* str)
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return FtFontGetStringWidthImpl(str) * (current->lineSpacing + current->maxHeight + current->heightOffset);
+}
+
+// 0x442278
+static int FtFontGetMonospacedCharacterWidthImpl()
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->lineSpacing + current->maxHeight;
+}
+                                                                                                                                                
+// 0x4422B4
+static void FtFontDrawImpl(unsigned char* buf, const char* string, int length, int pitch, int color)
+{
+    if (!gFtFontsInitialized) {
+        return;
+    }
+
+    if ((color & FONT_SHADOW) != 0) {
+        color &= ~FONT_SHADOW;
+        // NOTE: Other font options preserved. This is different from text font
+        // shadows.
+        FtFontDrawImpl(buf + pitch + 1, string, length, pitch, (color & ~0xFF) | _colorTable[0]);
+    }
+
+    unsigned char* palette = _getColorBlendTable(color & 0xFF);
```

Comment:
Building these blend tables are expensive. The GNW implementation I seen uses a reference counted lookup table and free up a particular blend table when the ref count reaches 0. In this context a blend table is built for `(color & ~0xFF) | _colorTable[0])` and another for `color & 0xFF`. As there are but a few text colors, like green, golden, etc. it would make sense to simply generate these blend tables for the most common text colors in the init phase and add reference counters to them. On exiting the font manager, `_freeColorBlendTable` could be called as many times as the text manager reference count value.